File: python/taflow/indicators/percent_above_moving_average.py
"""Python adapter for native Percent Above Moving Average breadth."""

from typing import Any
import numpy as np
from .._native import PercentAboveMovingAverage as _Native
from .._series import as_float64_series


class PercentAboveMovingAverage:
    """Return the percentage of constituents above a chosen moving average.

    Inputs are pre-aggregated counts; the caller chooses and evaluates the
    moving average per constituent. The result maps to Wickra ``PercentAboveMa``.

    Args:
        above_moving_average_count: Constituents above their reference average.
        universe_size: Total constituent count at each aligned tick.

    Raises:
        ValueError: If the two histories have different lengths.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native()

    def append(
        self, above_moving_average_count: float, universe_size: float
    ) -> "PercentAboveMovingAverage":
        """Append one market-wide count pair and return this adapter."""
        self._state.append(float(above_moving_average_count), float(universe_size))
        return self

    def extend(
        self, above_moving_average_count: Any, universe_size: Any
    ) -> "PercentAboveMovingAverage":
        """Append aligned count histories after length validation."""
        arrays = (
            as_float64_series(above_moving_average_count),
            as_float64_series(universe_size),
        )
        if len(arrays[0]) != len(arrays[1]):
            raise ValueError("above count and universe size must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest percentage, or ``None`` before input."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned percentage history."""
        return self._state.compute()

    def reset(self) -> "PercentAboveMovingAverage":
        """Reset native state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-tick count delegated to native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/percent_above_moving_average.rs
use crate::error::TaResult;
/// Percentage of constituents above their caller-selected moving average.
#[derive(Debug, Clone, Default)]
pub struct PercentAboveMovingAverage {
    count: usize,
    value: Option<f64>,
}
impl PercentAboveMovingAverage {
    /// Create an empty percentage state.
    pub fn new() -> TaResult<Self> {
        Ok(Self::default())
    }
    /// Append above-average and universe counts for one market tick.
    pub fn append(&mut self, above_moving_average_count: f64, universe_size: f64) -> Option<f64> {
        self.count += 1;
        self.value = Some(100.0 * above_moving_average_count / universe_size.max(1.0));
        self.value
    }
    /// Return the number of processed market ticks.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return the latest percentage above the moving average.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Restore fresh-state behaviour.
    pub fn reset(&mut self) {
        self.count = 0;
        self.value = None;
    }
}
